tests/scanner/test_scanner.py

```diff
@@ -294,6 +295,35 @@ async def test_invalid_media_path(self) -> None:
         with self.assertRaises(FileDirtyError):
             await self.scanner.scan_file(MEDIA_PATH + "/baz")
 
+    async def test_deduplicate_scans(self) -> None:
+        """Tests that if two scan requests come in for the same file and with the same
+        parameter, only one download/scan happens.
+        """
+
+        # Change the Mock's side effect to introduce some delay, to simulate a long
+        # download time. We sleep asynchronously to allow additional scans requests to be
+        # processed.
+        async def download_file(
+            media_path: str,
+            thumbnail_params: Optional[Dict[str, List[str]]] = None,
+        ) -> MediaDescription:
+            await asyncio.sleep(1.0)
```

Comment:
kind of unfortunate to introduce a real time-based sleep in the test. For one, it's possible for this to flake (though at 1 second probably unlikely). Otherwise, it's slowing down the test! :)

further I'm not sure we're even testing that you don't get two concurrent scans here — it appears that you `await` the first before the second returns.

I think I would use a Barrier: https://docs.python.org/3/library/asyncio-sync.html#barrier
instead of the sleep. Create a `Barrier(2)` and then `await barrier.wait()` instead of the sleep.
Instead of doing the requests one at once, you can `task1 = asyncio.create_task(self.scanner.scan_file(MEDIA_PATH))` (and etc for task2), then await them afterwards.

Reply:
TIL about Barrier, that sounds like what I'm after indeed!

Reply:
Oh but that won't work though, because Barriers are only a thing as of Python 3.11, and we want to test from 3.8.

Reply:
fair enough. If we make a pattern of adding more sleeps, we'll come up with a simple solution, but for now this is probably fine

Reply:
I'm not sure what this simple solution would be, since there isn't a method we can mock out that would be called by both `scan_file` calls (and I'd rather not add a function call to `scan_file` that we don't actually need, just so we can mock it out in tests). This would also be an issue if `Barriers` were a thing in Python < 3.11 fwiw.